routers/api/v1/repo/commits.go

```diff
@@ -120,3 +122,204 @@ func GetSingleCommit(ctx *context.APIContext) {
+	//   type: string
+	// - name: page
+	//   in: query
+	//   description: page number of requested commits
+	//   type: integer
+	// responses:
+	//   "200":
+	//     "$ref": "#/responses/CommitList"
+	//   "404":
+	//     "$ref": "#/responses/notFound"
+
+	gitRepo, err := git.OpenRepository(ctx.Repo.Repository.RepoPath())
+	if err != nil {
+		ctx.ServerError("OpenRepository", err)
+		return
+	}
+
+	page := ctx.QueryInt("page")
+	if page <= 0 {
+		page = 1
+	}
+
+	sha := ctx.Query("sha")
+
+	var baseCommit *git.Commit
+	if len(sha) == 0 {
+		// no sha supplied - use default branch
+		head, err := gitRepo.GetHEADBranch()
+		if err != nil {
+			ctx.ServerError("GetHEADBranch", err)
+			return
+		}
+		baseCommit, err = gitRepo.GetBranchCommit(head.Name)
+		if err != nil {
+			ctx.ServerError("GetCommit", err)
+			return
+		}
+	} else {
+		// get commit specified by sha
+		baseCommit, err = gitRepo.GetCommit(sha)
+		if err != nil {
+			ctx.ServerError("GetCommit", err)
+			return
+		}
+	}
+
+	// Total commit count
+	commitsCountTotal, err := baseCommit.CommitsCount()
+	if err != nil {
+		ctx.ServerError("GetCommitsCount", err)
+		return
+	}
+
+	pageCount := int(math.Ceil(float64(commitsCountTotal) / float64(git.CommitsRangeSize)))
+
+	// Query commits
+	commits, err := baseCommit.CommitsByRange(page)
+	if err != nil {
+		ctx.ServerError("CommitsByRange", err)
+		return
+	}
+
+	userCache := make(map[string]*models.User)
+
+	apiCommits := make([]*api.Commit, commits.Len())
+
+	i := 0
+	for commitPointer := commits.Front(); commitPointer != nil; commitPointer = commitPointer.Next() {
+		commit := commitPointer.Value.(*git.Commit)
+
+		var apiAuthor, apiCommitter *api.User
+
+		// Retrieve author and committer information
+		cacheAuthor, ok := userCache[commit.Author.Email]
+		if ok {
+			apiAuthor = cacheAuthor.APIFormat()
+		} else {
+			author, err := models.GetUserByEmail(commit.Author.Email)
+			if err != nil && !models.IsErrUserNotExist(err) {
+				ctx.ServerError("Get user by author email", err)
+				return
+			} else if err == nil {
+				apiAuthor = author.APIFormat()
+				userCache[commit.Author.Email] = author
+			}
+		}
+		cacheCommitter, ok := userCache[commit.Committer.Email]
+		if ok {
+			apiCommitter = cacheCommitter.APIFormat()
+		} else {
+			committer, err := models.GetUserByEmail(commit.Committer.Email)
+			if err != nil && !models.IsErrUserNotExist(err) {
+				ctx.ServerError("Get user by committer email", err)
+				return
+			} else if err == nil {
+				apiCommitter = committer.APIFormat()
+				userCache[commit.Committer.Email] = committer
+			}
+		}
+
+		// Retrieve parent(s) of the commit
+		apiParents := make([]*api.CommitMeta, commit.ParentCount())
+		for i := 0; i < commit.ParentCount(); i++ {
+			sha, _ := commit.ParentID(i)
+			apiParents[i] = &api.CommitMeta{
+				URL: ctx.Repo.Repository.APIURL() + "/git/commits/" + sha.String(),
+				SHA: sha.String(),
+			}
+		}
+
+		// Create json struct
+		apiCommits[i] = &api.Commit{
```

Comment:
This could be a sub-function similar to https://github.com/go-gitea/gitea/blob/37b6ee178705cd7e545ca3a8ae00b1f6cee89083/routers/api/v1/convert/convert.go#L50 That cloud be left in this file not necessarily to be put in convert module.
Furthermore, maybe de-duplicate with some parts with GetSingleCommit func.